barretenberg/cpp/src/barretenberg/stdlib/recursion/honk/verifier/ultra_recursive_verifier.cpp

```diff
@@ -191,8 +190,58 @@ std::array<typename Flavor::GroupElement, 2> UltraRecursiveVerifier_<Flavor, gob
          ")");
     prev_num_gates = builder->get_num_gates();
 
+    // Perform ECC op queue transcript aggregation protocol
+    if constexpr (IsGoblinFlavor<Flavor>) {
+        // Receive commitments [t_i^{shift}], [T_{i-1}], and [T_i]
+        std::array<Commitment, Flavor::NUM_WIRES> prev_agg_op_queue_commitments;
+        std::array<Commitment, Flavor::NUM_WIRES> shifted_op_wire_commitments;
+        std::array<Commitment, Flavor::NUM_WIRES> agg_op_queue_commitments;
+        for (size_t idx = 0; idx < Flavor::NUM_WIRES; ++idx) {
+            std::string suffix = std::to_string(idx + 1);
+            prev_agg_op_queue_commitments[idx] =
+                transcript.template receive_from_prover<Commitment>("PREV_AGG_OP_QUEUE_" + suffix);
+            shifted_op_wire_commitments[idx] =
+                transcript.template receive_from_prover<Commitment>("SHIFTED_OP_WIRE_" + suffix);
+            agg_op_queue_commitments[idx] =
+                transcript.template receive_from_prover<Commitment>("AGG_OP_QUEUE_" + suffix);
+        }
+
+        // Receive claimed evaluations of t_i^{shift}, T_{i-1}, and T_i
+        FF kappa = transcript.get_challenge("kappa");
```

Comment:
Don't you have to open one of these polynomials in gemini? Why not simply use rho?

Reply:
None of these polynomials are currently opened in gemini (only the unshifted t_i). Also, I believe the opening challenge must be produced after hashing the commitments to each of these polys so I think it would be awkward to try to squeeze that into Gemini.

Reply:
BTW we had discussed using the optimization of only committing to either t_i or its shift but not both, but it's not clear to me that its efficient in this case since we do not otherwise need to open both in Gemini. (For the other shifts, we already need to include them in Gemini since they're used in relations so there's no additional cost). Something we can continue to think about though.

Reply:
Not an issue right now, but you can commit to the polynomials at the same time as all the other wires (or right after). Then, when you receive the output from gemini, you ask for the evaluations for accumulated and previous versions (you have the shifted already).